src/lammpsio/snapshot.py

```diff
@@ -170,37 +258,84 @@ def to_hoomd_gsd(self, type_map=None):
             frame.particles.image = self.image.copy()
         if self.has_typeid():
             frame.particles.typeid = numpy.zeros(self.N, dtype=int)
-            if type_map is None:
-                sorted_typeids = numpy.sort(numpy.unique(self.typeid))
-                frame.particles.types = [str(typeid) for typeid in sorted_typeids]
-                for typeidx, typeid in enumerate(sorted_typeids):
-                    frame.particles.typeid[self.typeid == typeid] = typeidx
+            if type_map is not None:
+                warnings.warn(
+                    "type_map is deprecated, use label instead.",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
+                type_label_map = LabelMap(map=type_map)
             else:
-                frame.particles.types = list(type_map.values())
-                reverse_type_map = {
-                    typeid: typeidx for typeidx, typeid in enumerate(type_map.keys())
-                }
-                for i, typeid in enumerate(self.typeid):
-                    frame.particles.typeid[i] = reverse_type_map[typeid]
+                type_label_map = self.type_label
+            if self.has_typeid():
```

Comment:
These `if`s are a little off here (you have `self.has_typeid()` nested inside itself). I think you need to rejigger this a little to get the type label map before checking the typeid, more like how it's done below for the topology.